docs/guides/spmd_api.md

```diff
@@ -0,0 +1,432 @@
+from flax.core import meta
+from flax.core import freeze, unfreeze
+from flax.training import train_state, checkpoints
+
+import optax
+```
+
+<!-- #region id="XfuDTqt1mO6g" -->
+Import all the `pjit` related libraries. Note that they are still in the experimental package of JAX, so some API may change. In the near future, the plan is to move the `pjit` API out of the experimental status and merge it with the current `jax.jit` API.
+
+Start a device mesh using the 8 devices available. In this example, we set them as a `2x4` device mesh, and annotate each axis with a name. A typical way to annotate axis names is ('data', 'model'): 'data' being the mesh dimension used for data-parallel sharding of the batch dimension of inputs and activations, and the 'model' being the mesh dimension used for sharding parameters of the model across devices.
+<!-- #endregion -->
+
+```python executionInfo={"elapsed": 184, "status": "ok", "timestamp": 1671158220961, "user": {"displayName": "Ivy Zheng", "userId": "15297372265856137303"}, "user_tz": 480} id="vtDt-cA8u6sz" outputId="d3960c05-937c-4633-bb16-ab7c35d6899e"
+from jax.experimental.pjit import pjit, with_sharding_constraint, PartitionSpec
+from jax.experimental.maps import Mesh
+from jax.experimental import mesh_utils
+
+device_mesh = mesh_utils.create_device_mesh((2, 4))
+print(device_mesh)
+mesh = Mesh(devices=device_mesh, axis_names=('data', 'model'))
+mesh
+```
+
+<!-- #region id="SJcTXjGYhe5o" -->
+# Model Definition
+
+Define an example layer `DotReluDot`, which creates two parameters `W1` and `W2` for dot products, and does a `jax.nn.relu` in between. To use it in `pjit` efficiently, you can use APIs in `pjit` and `flax.core.meta` to annotate the parameters and intermediate variables correctly:
+
+* Use `meta.with_partitioning` to decorate the initializer function when creating params `W1` and `W2`.
+
+* Use `pjit.with_sharding_constraint` to annotate intermediate variables like `y` and `z` to force a particular sharding pattern under pjit when the ideal constraint is known.
+
+ * This is optional, but sometimes can help auto SPMD to partition efficiently.
+
+
+
+Each value passed into a sharding annotates a dimension, and its value should be among `data`, `model` or `None`. This refers to whether the data's axis should be sharded across one of the device mesh dimensions, or not sharded at all.
+<!-- #endregion -->
+
+```python executionInfo={"elapsed": 55, "status": "ok", "timestamp": 1671158221167, "user": {"displayName": "Ivy Zheng", "userId": "15297372265856137303"}, "user_tz": 480} id="YChzss7uwwHx"
+class DotReluDot(nn.Module):
+  depth: int
+  @nn.compact
+  def __call__(self, x):
+    W1 = self.param(
+        'W1', 
+        meta.with_partitioning(
+            nn.initializers.xavier_normal(),
+            (None, 'model')),
+        (x.shape[-1], self.depth))
+
+    y = jax.nn.relu(jnp.dot(x, W1))
+    # force a local sharding annotation, fairly redundant in this case.
```

Comment:
The local sharding annotation here. I removed the comment here and added more explanation on the documentation above. Basically `with_sharding_constraint` is only to ensure that `pjit` shards this variable in this way, and it's redundant here because `pjit` will by default shard it this way too.

Reply:
Ah makes sense, thanks for updating it!